src/components/NewClusterModal.js

```diff
@@ -285,16 +280,23 @@ export const NewClusterModal = withModalDrawer({ width: 675 })(class NewClusterM
   render() {
     const { currentCluster, onDismiss, onSuccess } = this.props
     const {
-      profile, masterMachineType, masterDiskSize, workerMachineType, numberOfWorkers, numberOfPreemptibleWorkers, workerDiskSize,
+      profile, masterMachineType, masterDiskSize, sparkMode, workerMachineType, numberOfWorkers, numberOfPreemptibleWorkers, workerDiskSize,
       jupyterUserScriptUri, selectedLeoImage, customEnvImage, leoImages, viewMode
     } = this.state
-    const { version, updated, packages } = _.find({ image: selectedLeoImage }, leoImages) || {}
+    const { version, updated, packages, requiresSpark } = _.find({ image: selectedLeoImage }, leoImages) || {}
+
+    const onEnvChange = ({ value }) => {
+      this.setState({
+        selectedLeoImage: value, customEnvImage: '',
+        sparkMode: _.find({ image: value }, leoImages)?.requiresSpark ? (sparkMode || 'master') : false
```

Comment:
Choosing Hail, selecting spark cluster, switching to non-Hail, and then back to Hail ends up in a state where 'spark master' is selected, but there are also non-zero workers. If we change `sparkMode` here, we also need to update `numWorkers`